le.go

```diff
@@ -186,31 +170,66 @@ func (logger *Logger) SetPrefix(prefix string) {
 // Write writes a bytes array to the Logentries TCP connection,
 // it adds the access token and prefix and also replaces
 // line breaks with the unicode \u2028 character
-func (logger *Logger) Write(p []byte) (n int, err error) {
-	if err := logger.ensureOpenConnection(); err != nil {
-		return 0, err
-	}
+func (logger *Logger) Write(p []byte) (int, error) {
+	fmt.Println("le_go: Write()")
 
-	logger.mu.Lock()
-	defer logger.mu.Unlock()
+	buf := logger.makeBuf(p)
 
-	logger.makeBuf(p)
+	n, err := logger.conn.Write(buf)
+	if err == nil {
+		return n, err
+	}
+	fmt.Printf("le_go: Write n=%d, err=%s", n, err.Error())
 
-	return logger.conn.Write(logger.buf)
+	// First write failed.  Try reconnecting and then a second write; if that fails give up.  If
+	// we wanted to keep trying we would have to maintain a queue and a separate goroutine.
+	if err = logger.reopenConnection(); err != nil {
+		return 0, err
+	}
+	n, err = logger.conn.Write(buf)
+	if err != nil {
+		fmt.Printf("le_go: Write n=%d, err=%s", n, err.Error())
+	}
+	return n, err
+}
+
+// bytes.IndexByte exists but not bytes.CountByte
+func countByte(s []byte, c byte) int {
+	return bytes.Count(s, []byte{c})
 }
 
 // makeBuf constructs the logger buffer
-// it is not safe to be used from within multiple concurrent goroutines
-func (logger *Logger) makeBuf(p []byte) {
-	count := strings.Count(string(p), lineSep)
-	p = []byte(strings.Replace(string(p), lineSep, "\u2028", count-1))
-
-	logger.buf = logger.buf[:0]
-	logger.buf = append(logger.buf, (logger.token + " ")...)
-	logger.buf = append(logger.buf, (logger.prefix + " ")...)
-	logger.buf = append(logger.buf, p...)
-
-	if !strings.HasSuffix(string(logger.buf), lineSep) {
-		logger.buf = append(logger.buf, (lineSep)...)
+func (logger *Logger) makeBuf(p []byte) []byte {
+	// Pre-allocate a buffer of the correct size
+	capacity := len(logger.token) + 1
+	capacity += len(logger.prefix) + 1
+	capacity += len(p)                         // nominal payload size (before replacement)
+	capacity += countByte(p, asciiLineSep) * 2 // 1-byte "\n"s replaced with 3-byte "\u2028"s
```

Comment:
Yeah, it did, but the Python version of this client doesn't bother: it replaces all newlines, then adds a `\n` at the end, unconditionally.